File: app/src/main/java/de/hype/hypenotify/core/BackgroundService.java
package de.hype.hypenotify.core;

import android.annotation.SuppressLint;
import android.content.Intent;
import androidx.core.app.NotificationCompat;
import de.hype.hypenotify.MainActivity;
import de.hype.hypenotify.NotificationUtils;
import de.hype.hypenotify.PermissionUtils;
import de.hype.hypenotify.R;
import de.hype.hypenotify.core.interfaces.Core;
import de.hype.hypenotify.services.HypeNotifyService;
import de.hype.hypenotify.tools.notification.Notification;
import de.hype.hypenotify.tools.notification.NotificationBuilder;
import de.hype.hypenotify.tools.notification.NotificationChannels;
import org.jetbrains.annotations.Blocking;

import java.util.function.Consumer;

public class BackgroundService extends HypeNotifyService<BackgroundService> {
    @SuppressLint("StaticFieldLeak")
    public static BackgroundService instance;
    private static boolean isRunning = false;
    private MiniCore core;

    /**
     * Make sure you start the Service before calling this method!
     */
    @Blocking
    public static BackgroundService getInstanceBlocking() {
        while (instance == null) {
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
        return instance;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        NotificationUtils.synchronizeNotificationChannels(this);
        NotificationBuilder notificationBuilder = new NotificationBuilder(this, "Background", "A HypeNotify Service is running in the background.", NotificationChannels.BACKGROUND_SERVICE);
        notificationBuilder.setSmallIcon(R.mipmap.icon);
        notificationBuilder.setLargeImage(R.mipmap.icon);
        notificationBuilder.getHiddenBuilder().setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_DEFERRED);
        Notification notification = notificationBuilder.build();
        startForeground(notification.getID(), notification.get());
        startCore();
        instance = this;
    }

    private void startCore() {
        // Start your service and perform initialization.
        core = new ExpandedMiniCore(this);

        if (!PermissionUtils.checkPermissions(this)) {
            NotificationBuilder notificationBuilder = new NotificationBuilder(
                    this,
                    "Permissions Missing",
                    "We noticed that HypeNotify is lacking Permissions. Please open the app and grant the permissions.",
                    NotificationChannels.ERROR
            );
            notificationBuilder.send();
        }

        core.executionService.execute(() -> {
            try {
                core.fullInit();
            } catch (Exception e) {
                e.printStackTrace();
            }

            if (!core.areKeysSet()) {
                NotificationBuilder notificationBuilder = new NotificationBuilder(
                        this,
                        "Keys Missing",
                        "We noticed that HypeNotify is lacking Keys. Please open the app and enter your keys.",
                        NotificationChannels.ERROR
                );
                notificationBuilder.send();
            }
        });
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (isRunning) {
            return START_NOT_STICKY;
        }
        isRunning = true;
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        core.onDestroy();
        isRunning = false;
        instance = null;
    }

    /**
     * Will wait until
     *
     * @param consumer consumer to be run.
     */
    public static void executeWithBackgroundService(Consumer<BackgroundService> consumer) {
        Thread thread = new Thread(() -> {
            BackgroundService instance = getInstanceBlocking();
            consumer.accept(instance);
        });
        thread.start();
    }

    public Core getCore(MainActivity context) {
        return new CoreImpl(context, core);
    }

    @Override
    public de.hype.hypenotify.core.interfaces.MiniCore getCore() {
        return core;
    }

}

File: app/src/main/java/de/hype/hypenotify/app/core/BackgroundService.kt
package de.hype.hypenotify.app.core

import android.annotation.SuppressLint
import android.app.AlarmManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.SystemClock
import androidx.core.app.NotificationCompat
import de.hype.hypenotify.R
import de.hype.hypenotify.app.MainActivity
import de.hype.hypenotify.app.core.interfaces.Core
import de.hype.hypenotify.app.core.interfaces.MiniCore
import de.hype.hypenotify.app.services.HypeNotifyService
import de.hype.hypenotify.app.tools.notification.NotificationBuilder
import de.hype.hypenotify.app.tools.notification.NotificationChannels
import org.jetbrains.annotations.Blocking
import java.util.function.Consumer

class BackgroundService : HypeNotifyService<BackgroundService>() {
    override var core: MiniCore

    private val closeReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context?, i: Intent?) {
            stopSelf()
        }
    }
    private val dismissReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context?, i: Intent?) {
            showNotificationWithCloseButton()
        }
    }

    override fun onCreate() {
        super.onCreate()

        // Register receivers
        registerReceiver(
            closeReceiver,
            IntentFilter(CLOSE_SERVICE_ACTION),
            RECEIVER_NOT_EXPORTED
        )
        registerReceiver(
            dismissReceiver,
            IntentFilter(DISMISS_ACTION),
            RECEIVER_NOT_EXPORTED
        )

        showNotificationWithCloseButton()
        instance = this
        core = ExpandedMiniCore(this)
    }

    private fun showNotificationWithCloseButton() {
        val closeIntent: Intent? = Intent(CLOSE_SERVICE_ACTION)
            .setPackage(getPackageName())
        val closePI = PendingIntent.getBroadcast(
            this, 0, closeIntent!!, PendingIntent.FLAG_IMMUTABLE
        )

        val deleteIntent: Intent? = Intent(DISMISS_ACTION)
            .setPackage(getPackageName())
        val deletePI = PendingIntent.getBroadcast(
            this, 0, deleteIntent!!, PendingIntent.FLAG_IMMUTABLE
        )

        val nb = NotificationBuilder(
            this,
            "Background",
            "A HypeNotify Service is running in the background.",
            NotificationChannels.BACKGROUND_SERVICE
        )

        nb.setSmallIcon(R.mipmap.icon)
        nb.setLargeImage(R.mipmap.icon)
        nb.hiddenBuilder
            .setOngoing(true)
            .setForegroundServiceBehavior(
                NotificationCompat.FOREGROUND_SERVICE_DEFERRED
            )
            .addAction(
                android.R.drawable.ic_menu_close_clear_cancel,
                "Stop Service",
                closePI
            )
            .setDeleteIntent(deletePI)

        startForeground(nb.build().iD, nb.build().get())
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (isRunning) return START_NOT_STICKY
        isRunning = true
        return START_STICKY
    }

    override fun onTaskRemoved(rootIntent: Intent?) {
        // If we previously stopped service due to low battery, don't restart automatically.
        if (core != null && core.isLowBatteryStop()) {
            super.onTaskRemoved(rootIntent)
            return
        }

        val restart: Intent? = Intent(
            getApplicationContext(),
            BackgroundService::class.java
        )
            .setPackage(getPackageName())
        val pi = PendingIntent.getService(
            getApplicationContext(), 1, restart!!,
            PendingIntent.FLAG_IMMUTABLE
        )
        val am = getSystemService(
            ALARM_SERVICE
        ) as AlarmManager
        am.set(
            AlarmManager.ELAPSED_REALTIME,
            SystemClock.elapsedRealtime() + 1000, pi!!
        )
        super.onTaskRemoved(rootIntent)
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(closeReceiver)
        unregisterReceiver(dismissReceiver)
        instance = null
        isRunning = false
    }


    @Blocking
    fun getCore(context: MainActivity): Core {
        while (core == null) {
            try {
                Thread.sleep(200)
            } catch (e: InterruptedException) {
            }
        }
        return CoreImpl(context, core)
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        var instance: BackgroundService? = null
        private var isRunning = false
        private const val CLOSE_SERVICE_ACTION = "de.hype.hypenotify.CLOSE_SERVICE_ACTION"
        private const val DISMISS_ACTION = "de.hype.hypenotify.NOTIF_DISMISSED"

        @get:Blocking val instanceBlocking: BackgroundService
            /**
             * Make sure you start the Service before calling this method!
             */
            get() {
                while (instance == null) {
                    try {
                        Thread.sleep(500)
                    } catch (e: InterruptedException) {
                        throw RuntimeException(e)
                    }
                }
                return instance!!
            }

        /**
         * Will wait until
         *
         * @param consumer consumer to be run.
         */
        fun executeWithBackgroundService(consumer: Consumer<BackgroundService?>) {
            val thread = Thread(Runnable {
                val instance: BackgroundService = instanceBlocking
                consumer.accept(instance)
            })
            thread.start()
        }
    }
}
